src/common/tracer.h

```diff
@@ -46,42 +46,80 @@ class Tracer {
 
 };
 
-inline void encode(const jspan_context& span_ctx, bufferlist& bl, uint64_t f = 0) {
-  ENCODE_START(1, 1, bl);
+// encoding the span
+// shoudl be called only if valid, and between ENCODE_START, ENCODE_FINISH pair
+inline void encode_nohead(const jspan_context& span_ctx, bufferlist& bl, uint64_t f = 0) {
+  using namespace opentelemetry;
+  using namespace trace;
+  ceph::encode_nohead(std::string_view(reinterpret_cast<const char*>(span_ctx.trace_id().Id().data()), TraceId::kSize), bl);
+  ceph::encode_nohead(std::string_view(reinterpret_cast<const char*>(span_ctx.span_id().Id().data()), SpanId::kSize), bl);
+  encode(span_ctx.trace_flags().flags(), bl);
+}
+
+// decoding the span
+// shoudl be called only if valid, and between DECODE_START, DECODE_FINISH pair
+inline void decode_nohead(jspan_context& span_ctx, bufferlist::const_iterator& bl) {
   using namespace opentelemetry;
   using namespace trace;
+  std::array<uint8_t, TraceId::kSize> trace_id;
+  std::array<uint8_t, SpanId::kSize> span_id;
+  uint8_t flags;
+  decode(trace_id, bl);
+  decode(span_id, bl);
+  decode(flags, bl);
+  span_ctx = SpanContext(
+    TraceId(nostd::span<uint8_t, TraceId::kSize>(trace_id)),
+    SpanId(nostd::span<uint8_t, SpanId::kSize>(span_id)),
+    TraceFlags(flags),
+    true);
+}
+
+// encoding the span
+inline void encode(const jspan_context& span_ctx, bufferlist& bl, uint64_t f = 0) {
+  ENCODE_START(1, 1, bl);
   auto is_valid = span_ctx.IsValid();
   encode(is_valid, bl);
   if (is_valid) {
-    encode_nohead(std::string_view(reinterpret_cast<const char*>(span_ctx.trace_id().Id().data()), TraceId::kSize), bl);
-    encode_nohead(std::string_view(reinterpret_cast<const char*>(span_ctx.span_id().Id().data()), SpanId::kSize), bl);
-    encode(span_ctx.trace_flags().flags(), bl);
+    encode_nohead(span_ctx, bl, f);
   }
   ENCODE_FINISH(bl);
 }
 
+// decoding the span
 inline void decode(jspan_context& span_ctx, bufferlist::const_iterator& bl) {
-  using namespace opentelemetry;
-  using namespace trace;
   DECODE_START(1, bl);
   bool is_valid;
   decode(is_valid, bl);
   if (is_valid) {
-    std::array<uint8_t, TraceId::kSize> trace_id;
-    std::array<uint8_t, SpanId::kSize> span_id;
-    uint8_t flags;
-    decode(trace_id, bl);
-    decode(span_id, bl);
-    decode(flags, bl);
-    span_ctx = SpanContext(
-      TraceId(nostd::span<uint8_t, TraceId::kSize>(trace_id)),
-      SpanId(nostd::span<uint8_t, SpanId::kSize>(span_id)),
-      TraceFlags(flags),
-      true);
+    decode_nohead(span_ctx, bl);
   }
   DECODE_FINISH(bl);
 }
 
+// encoding the span
+// shoudl be called between ENCODE_START, ENCODE_FINISH pair
+inline void nested_encode(const jspan_context& span_ctx, bufferlist& bl, uint64_t f = 0) {
+  auto is_valid = span_ctx.IsValid();
+  ceph::encode(is_valid, bl);
```

Comment:
Doesn't this change break the existing encoding?

Reply:
@athanatos this specific change is part of an optimization to avoid parsing a struct when encoding is disabled (by taking the boolean out of the struct). see: ea8f7fb7118de15d55dace1031fbe7d82809e1b4

but regardless of that optimization (that we can remove), adding the new trace struct in the middle of the message, breaks existing encoding. 
to solve this issue I'm wating for "squid" feature bit to be merged: https://github.com/ceph/ceph/pull/53191